<img style="float: left; margin: 30px 15px 15px 15px;" src="https://pngimage.net/wp-content/uploads/2018/06/logo-iteso-png-5.png" width="300" height="500" /> 
    
    
### <font color='navy'> Simulación de procesos financieros. 

**Nombres:** Alejandra Rico Rosales, Diego Alberto López Lazareno.

**Fecha:** 22 de octubre del 2021.

**Expediente:** 722100, 714643.
    
**Profesor:** Oscar David Jaramillo Zuluaga.
    
**Link Github:** [Link con el enlace del repositorio del creador del proyecto](https://github.com/alerrico/Tarea8_ARico_DALopez)

# Tarea 8: Clase 14

In [1]:
import numpy as np
import scipy.stats as st
import pandas as pd
import matplotlib.pyplot as plt

## Enunciado de tarea (Transcribirlo)

<img src="tarea.jpg">

> Además use el método de estratificación donde se divide en B estratos, visto la clase pasada y compare el resultado con los métodos anteriores cuando se toman $2,4,6,10$ estratos respectivamente, concluya.

#### Código de solución `ESTUDIANTE 1: ALEJANDRA RICO`

$$\int_{-1}^{x}\frac{3x^2}{2}dx=\left(\frac{3}{2}\right)\left(\frac{x^3}{3}\right)|_{-1}^x=\frac{x^3}{2}+\frac{1}{2}$$

Igualamos a u:
$$\frac{x^3}{2}+\frac{1}{2}=u$$

Despejamos x:
$$x=\sqrt[3]{2u-1}$$

In [23]:
# Montecarlo crudo

N = 1000

f = lambda x: (3/2)*x**2 if -1<=x<=1 else 0
h = lambda u: (2*u-1)**(1/3) if u > 0.5 else -(1-2*u)**(1/3)

u1 = np.random.rand(N)
fu1 = list(map(f,u1))
m_montecarlo = np.mean(list(map(h, u1)))

print('Montecarlo crudo: ', m_montecarlo)

Montecarlo crudo:  0.020783088098140733


In [24]:
# Método estratificado 
r1 = np.random.uniform(0,0.6,3)
r2 = np.random.uniform(0.6,0.9,3)
r3 = np.random.uniform(0.9,1,4)
r = [r1,r2,r3]
w = [1/2,1,4]


f_inv = lambda x: np.array(list(map(lambda i: h(i),x)))
xi = list(map(f_inv,r))
m = np.concatenate(list(map(lambda i,wi: i/wi,xi,w)))
me = m.mean()

print('Método estratificado: ', me)

Método estratificado:  0.19957728718642087


In [25]:
# Complementarios

u2 = np.random.rand(int(N/2))
u2_c = 1 - u2

m_comp = np.mean(list(map(lambda u: h(u), np.concatenate([u2, u2_c]))))
m_comp

print('Complementarios: ', m_comp)

Complementarios:  1.7763568394002505e-18


In [27]:
# Método de estratificación dividido en B estratos

def estratificado(N: "Numero de muestras"):
    U2 = np.random.rand(N)
    i = np.arange(0,N)
    return (U2+i)/N

B = np.array([2, 4, 6, 10])
fitt = [estratificado(bi) for bi in B]
estrato = [[h(var) for var in x] for x in fitt]

estratos = list([np.array(estrato[i]).mean() for i in range(4)])
print('Con 2, 4, 6 y 10 estratos: ', estratos)

Con 2, 4, 6 y 10 estratos:  [0.0579630460498593, 0.004993531412614993, 0.0025720692712120443, 0.007466393735603638]


#### Código de solución `ESTUDIANTE 2: DIEGO LOPEZ`

$$ f(x) = \frac{3}{2} x^2, -1 \leq x \leq 1 $$ 

#### Valor Esperado

Se obtiene el valor esperado de la Función de Densidad de Probabilidad analíticamente:

$$ E[X] = \int^\infty_{-\infty} x f_X(x) dx = \int^1_{-1} x \frac{3}{2} \space x^2 dx = \frac{3}{2} \int^1_{-1} x^3 dx = \frac{3}{8} x^4 \Big|^1_{-1} = \frac{3}{8} (1)^4 - \frac{3}{8} (-1)^4 = \frac{3}{8} - \frac{3}{8} = 0 $$

#### CDF

Se obtiene la Función de Distribución Acumulada:

$$ F(x) = \int^\infty_{-\infty} f(x) dx = \int^x_{-1} \frac{3}{2} \space x^2 dx = \frac{1}{2} x^3 \Big|^x_{-1} = \frac{1}{2} x^3 - \frac{1}{2} (-1)^3 = \frac{1}{2} x^3 + \frac{1}{2} = \frac{1}{2} (x^3 + 1) $$

#### Método de la Transformada Inversa

$$ F(x) = \frac{1}{2} (x^3 + 1) $$

$$ U \sim [0, 1] = \frac{1}{2} (x^3 + 1) $$

$$ U = \frac{1}{2} (x^3 + 1) $$

$$ x = (2 U - 1)^{1/3} $$

para el intervalo:

$$ 0 \leq U \leq 1 $$ 


In [2]:
# Configuraciones iniciales: 10 números aleatorios uniformemente distribuidos entre 0 y 1
np.random.seed(101)
N = 10
U = np.random.rand(N)

In [3]:
# (a) Método de MonteCarlo Crudo
transformadaInversa = lambda U : (2 * U - 1) ** (1 / 3) if U >= 0.5 else - (-2 * U + 1) ** (1 / 3)
x_a = [transformadaInversa(u) for u in U]

print("La media de la distribución, por el Método de MonteCarlo Crudo, resulta :", np.mean(x_a))

La media de la distribución, por el Método de MonteCarlo Crudo, resulta : 0.0688170177133047


In [4]:
# (b) Método de Muestreo Estratificado
# Estratos
e = [(0, 0.6), (0.6, 0.9), (0.9, 1)]
# Números aleatorios por estrato
muestras = [np.random.uniform(e[0][0], e[0][1], 3), np.random.uniform(e[1][0], e[1][1], 3), 
            np.random.uniform(e[2][0], e[2][1], 4)]
# Pesos
pesos = [(len(muestras[i]) / len(np.concatenate(muestras))) / (e[i][1] - e[i][0]) for i in range(len(muestras))]

# Muestras ponderadas
xi = [[transformadaInversa(j) for j in i] for i in muestras]
x_b = np.concatenate([[xi[i][j] / pesos[i] for j in range(len(xi[i]))] for i in range(len(xi))])

print("La media de la distribución, por el Método de Muestreo Estratificado, resulta :", np.mean(x_b))

La media de la distribución, por el Método de Muestreo Estratificado, resulta : -0.14437648350109505


In [5]:
# (c) Método de los Números Aleatorios Complementarios
x_c = [transformadaInversa(u) for u in list(U[0 : int(N / 2)]) + list(1 - U[0 : int(N / 2)])]

print("La media de la distribución, por el Método de los Números Aleatorios Complementarios, resulta :", np.mean(x_c))

La media de la distribución, por el Método de los Números Aleatorios Complementarios, resulta : -1.1102230246251566e-17


In [6]:
# (d) Método de Muestreo Estratificado con 2, 4, 6 y 10 estratos

def muestreoEstratificado(B : "Cantidad de estratos"):
    return (np.random.rand(B) + np.arange(0, B)) / B

estratos = [2, 4, 6, 10]
x_d = [[transformadaInversa(u) for u in muestreoEstratificado(B)] for B in estratos]

[print("La media de la distribución, cuando se tienen " + str(B) + " estratos, resulta :", 
       np.mean(x_d[i])) for i, B in zip(range(len(estratos)), estratos)];

La media de la distribución, cuando se tienen 2 estratos, resulta : 0.061427897906401074
La media de la distribución, cuando se tienen 4 estratos, resulta : -0.07358960333651912
La media de la distribución, cuando se tienen 6 estratos, resulta : 0.04013182147565967
La media de la distribución, cuando se tienen 10 estratos, resulta : 0.024286171468082252


El valor esperado de la PDF es $0$. Observamos que el método que más se acerca a este resultado es el de los **Números Aleatorios Complementarios**, dando una media de $-1.11 e-17$. Seguido de este método tenemos que el **Muestreo Estratificado** con $6$ y $10$ estratos arroja resultados decentes, de $0.04$ y $0.02$ respectivamente; sin embargo, cuando únicamente se tienen $2$ estratos la media aproximada resulta peor que en el caso del **MonteCarlo Crudo**, por lo que, si se quieren resultados fiables, lo más recomendable es emplear más estratos. En el caso del **Método de Muestreo Estratificado** con los intervalos propuestos se llega a resultados bastante dispersos. A manera general podemos afirmar que los métodos de reducción de la varianza, en su mayoría, son mejores técnicas que el **MonteCarlo Crudo.**